policy.go

```diff
@@ -784,95 +960,118 @@ func executePolicySession(tpm *tpm2.TPMContext, policySession tpm2.SessionContex
+			return staticPolicyDataError{errors.New("public area of dynamic authorization policy signing key is invalid")}
 		}
-		return xerrors.Errorf("cannot load public area for dynamic authorization policy signature verification key: %w", err)
+		return xerrors.Errorf("cannot load public area for dynamic authorization policy signing key: %w", err)
 	}
 	defer tpm.FlushContext(authorizeKey)
 
+	var pcrPolicyRef tpm2.Nonce
+	if version > 0 {
+		pcrPolicyRef = computePcrPolicyRefFromCounterContext(policyCounter)
+	}
+
 	h := authPublicKey.NameAlg.NewHash()
-	h.Write(dynamicInput.AuthorizedPolicy)
+	h.Write(dynamicInput.authorizedPolicy)
+	h.Write(pcrPolicyRef)
 
-	authorizeTicket, err := tpm.VerifySignature(authorizeKey, h.Sum(nil), dynamicInput.AuthorizedPolicySignature)
+	authorizeTicket, err := tpm.VerifySignature(authorizeKey, h.Sum(nil), dynamicInput.authorizedPolicySignature)
 	if err != nil {
 		if tpm2.IsTPMParameterError(err, tpm2.AnyErrorCode, tpm2.CommandVerifySignature, 2) {
-			// dynamicInput.AuthorizedPolicySignature is invalid.
-			return dynamicPolicyDataError{errors.New("cannot verify dynamic authorization policy signature")}
+			// dynamicInput.AuthorizedPolicySignature or the computed policy ref is invalid.
+			// XXX: It's not possible to determine whether this is broken dynamic or static metadata -
+			//  we should just do away with the distinction here tbh
+			return dynamicPolicyDataError{errors.New("cannot verify PCR policy signature")}
 		}
-		return xerrors.Errorf("cannot verify dynamic authorization policy signature: %w", err)
+		return xerrors.Errorf("cannot verify PCR policy signature: %w", err)
 	}
 
-	if err := tpm.PolicyAuthorize(policySession, dynamicInput.AuthorizedPolicy, nil, authorizeKey.Name(), authorizeTicket); err != nil {
+	if err := tpm.PolicyAuthorize(policySession, dynamicInput.authorizedPolicy, pcrPolicyRef, authorizeKey.Name(), authorizeTicket); err != nil {
 		if tpm2.IsTPMParameterError(err, tpm2.ErrorValue, tpm2.CommandPolicyAuthorize, 1) {
 			// dynamicInput.AuthorizedPolicy is invalid.
-			return dynamicPolicyDataError{errors.New("the dynamic authorization policy is invalid")}
+			return dynamicPolicyDataError{errors.New("the PCR policy is invalid")}
 		}
-		return xerrors.Errorf("dynamic authorization policy check failed: %w", err)
+		return xerrors.Errorf("PCR policy check failed: %w", err)
 	}
 
-	pinIndex.SetAuthValue([]byte(pin))
-	if _, _, err := tpm.PolicySecret(pinIndex, policySession, nil, nil, 0, hmacSession); err != nil {
-		return xerrors.Errorf("cannot execute PolicySecret assertion: %w", err)
+	if version == 0 {
+		policyCounter.SetAuthValue([]byte(pin))
+		if _, _, err := tpm.PolicySecret(policyCounter, policySession, nil, nil, 0, hmacSession); err != nil {
+			return xerrors.Errorf("cannot execute PolicySecret assertion: %w", err)
+		}
+	} else {
+		if err := tpm.PolicyAuthValue(policySession); err != nil {
```

Comment:
this corresponds to https://github.com/snapcore/secboot/pull/92/files#diff-9c53128bbf22cd8275aacdbf0ae82035R726 ?

Reply:
I've added a comment here.